sphinx/source/docs/dev_guide/bokehjs.rst

```diff
@@ -161,41 +178,133 @@ you can run tests immediately with these endpoints:
-   test run for your PR and download the associated ``bokehjs-report`` artifact.
-7. Unzip the artifact archive at the root of the repository.
-8. Assuming devtools server is running in the background, go to ``/integration/report?platform=name``
-   where ``name`` is either ``linux``, ``macos`` or ``windows`` and review the test output
-   for each platform. If there are no unintentional differences, then commit all
-   new or modified ``*.blf`` and ``*.png`` files under ``test/baselines/{linux,macos,windows}``.
-9. Push your changes to GitHub again and verify that tests pass this time.
+BokehJS uses a series of visual baseline comparison tests. These tests help make sure
+that Bokeh's visual output is consistent with the output expected by design.
+
+In the background, BokehJS' testing framework runs a headless browser and takes
+screenshots of the browser's output. The testing framework then compares the visual
+output to each test's dedicated baseline files.
+
+Each test in ``test:integration`` consists of two types of baseline comparison:
+
+Textual baseline comparison
+  For each test, the testing framework compares the pixel location of certain elements
+  in the visual output to pixel locations in a baseline file. The pixel locations are a
+  set of bounding boxes. They are stored as plain text in each test's ``.blf`` files.
+
+Visual baseline comparison
+  For each test, the testing suite does a pixel-by-pixel comparison of a screenshot
+  and a baseline image. These baseline images are stored as ``.png`` files. In contrast
+  to textual baseline comparisons, visual baseline comparisons are platform-dependent.
+  Even minor differences in font rendering, for example, will make the pixel-by-pixel
+  comparison fail.
+
+The visual baseline comparison tests are located in the
+:bokeh-tree:`bokehjs/test/integration/` folder and sub-folders.
+:ref:`Bokeh's CI <devguide_testing_ci>` runs these tests on Linux, macOS, and Windows
+environments. The baseline files for each environment are located in the
+:bokeh-tree:`bokehjs/test/baselines/` folder.
+
+Follow these steps to write new visual tests or update existing tests:
+
+1. Create or update visual testing scripts:
+    To write a visual test for BokehJS' testing suite, start by importing the
+    ``display()`` and ``fig()`` functions from the testing framework's ``_util`` module:
+
+    .. code-block:: TypeScript
+
+        import {display, fig} from "../_util"
+
+    The ``display()`` function in ``_util`` is similar to BokehJS' standard ``show()``
+    function: ``display()`` accepts the same arguments as ``show()`` but also captures
+    the visual output for comparison.
+
+    The ``fig()`` function in ``_util`` is similar to BokehJS' standard ``figure()``
+    function: ``fig()`` accepts the same arguments as ``figure()``. However, to keep
+    visual tests as efficient as possible, you should only use ``width`` and ``height``,
+    if possible.
```

Comment:
@mattpap Is that a somewhat accurate summary of ``display()`` and ``fig()`` to communicate the most important characteristics in this context?

Reply:
a important difference here is that ``fig()`` takes ``figure()`` attributes as it's second parameter, note:
```typescript
// bokehjs/src/lib/api/plotting.ts
figure(attributes?: Partial<Figure.Attrs>)
```
```typescript
// bokehjs/test/integration/_util.ts
fig([width, height]: [number, number], attrs?: Partial<Figure.Attrs>)
```

Reply:
Thanks @IuryPiva, I have updated the text to reflect that!